src/Ormolu/Terminal.hs

```diff
@@ -17,114 +17,121 @@ module Ormolu.Terminal
 
     -- * Printing
     put,
-    putS,
-    putSrcSpan,
-    putRealSrcSpan,
+    putShow,
+    putOutputable,
     newline,
   )
 where
 
 import Control.Monad.Reader
 import Data.Text (Text)
+import qualified Data.Text as T
 import qualified Data.Text.IO as T
-import GHC.Types.SrcLoc
+import GHC.Utils.Outputable (Outputable)
 import Ormolu.Utils (showOutputable)
 import System.Console.ANSI
-import System.IO (Handle, hFlush, hPutStr)
+import System.IO (Handle, hFlush)
 
 ----------------------------------------------------------------------------
 -- The 'Term' monad
 
--- | Terminal monad.
-newtype Term a = Term (ReaderT RC IO a)
-  deriving (Functor, Applicative, Monad)
+type Term = TermOutput ()
 
--- | Reader context of 'Term'.
-data RC = RC
-  { -- | Whether to use colors
-    rcUseColor :: Bool,
-    -- | Handle to print to
-    rcHandle :: Handle
-  }
+newtype TermOutput a = TermOutput [TermOutputNode]
+
+data TermOutputNode
+  = OutputText Text
+  | WithColor Color Term
+  | WithBold Term
+
+instance Functor TermOutput where
+  fmap _ (TermOutput nodes) = TermOutput nodes
+
+instance Applicative TermOutput where
+  pure _ = TermOutput []
+  TermOutput nodes1 <*> TermOutput nodes2 = TermOutput (nodes1 <> nodes2)
+
+instance Monad TermOutput where
+  TermOutput nodes1 >>= k =
+    let TermOutput nodes2 = k (error "TermOutput unexpectedly evaluated continuation")
+     in TermOutput (nodes1 <> nodes2)
```

Comment:
If you want, I think we could use `QualifiedDo` here to avoid these somewhat weird instances as you note, i.e. make `TermOutput` monomorphic, `newtype`-derive `Monoid`, define
```haskell
(>>) :: TermOutput -> TermOutput
(>>) = (<>)
```
and then make the call sites qualified and use `Term.do` instead of `do`.

Also see the ["Monoids" example](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0216-qualified-do.rst#examples) in the original proposal.

Reply:
Nice! I went even further and removed the Applicative instance (cause I still wasn't crazy about the Functor/Applicative instances). I'm less happy about the fact that I needed to implement new `for'`, `when'`, and `unless'` combinators, but I do like the removal of the phantom type parameter. Thoughts?

Reply:
I think I might very slightly prefer `Applicative` as one can then reuse `for_`/`when`/`unless`, but feel free to leave it as is :+1:

Also note that `for' = flip foldMap`, so it could probably be inlined and would not need a dedicated binding.

Reply:
You can even derive the `Applicative` using `DerivingVia`:
```haskell
newtype Term' a = Term (DList TermOutputNode)
  deriving (Functor, Applicative) via Const [TermOutputNode]
```